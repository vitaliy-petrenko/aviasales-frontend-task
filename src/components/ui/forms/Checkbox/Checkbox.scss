@import '../../../../styles/abstract';

.checkbox {
  display: flex;
  align-items: flex-start;
  cursor: pointer;
  user-select: none;

  input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;

    &:checked ~ .checkbox__icon {
      border-color: var(--color-form-option-active);

      svg {
        opacity: 1;
      }
    }

    &:focus ~ .checkbox__icon {
      box-shadow: var(--default-focus-shadow);
    }
  }
}

.checkbox__icon {
  display: block;
  width: 20px;
  height: 20px;
  border: 1px solid var(--color-form-option);
  border-radius: 2px;
  transition: border-color .05s;
  position: relative;
  margin-right: $grid-base/2;

  svg {
    position: absolute;
    opacity: 0;
    transition: opacity .05s;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);

    path {
      fill: var(--color-form-option-active)
    }
  }
}

.checkbox__label {
  @include font-size-base;
  @include word-break;
}
