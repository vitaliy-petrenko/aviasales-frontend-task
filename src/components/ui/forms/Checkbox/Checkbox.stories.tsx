import React, { useState } from 'react'
import Checkbox from './Checkbox'
import { storiesOf } from '@storybook/react'
import results from '../../../../../.jest-test-results.json'
import { withTests } from '@storybook/addon-jest'

const params = {
  jest: ['ButtonGroup.spec.tsx'],
}

storiesOf('Checkbox', module)
  .addDecorator(withTests({ results }))
  .add('Checkbox', () => {
      const
        [checked, setChecked] = useState(true),
        onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
          setChecked(event.target.checked)
        }

      return (
        <Checkbox checked={checked} onChange={onChange}>label</Checkbox>
      )
    }, params
  )

