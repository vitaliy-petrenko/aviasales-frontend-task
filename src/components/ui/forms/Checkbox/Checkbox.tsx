import React from 'react'
import { ReactComponent as ICheck } from '../../../../assets/icons/check.svg'
import './Checkbox.scss'

interface IProps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'onChange'> {
  onChange(value: boolean): void

  type?: string
}

const Checkbox: React.FC<IProps> = ({ children, onChange, type = 'checkbox', ...rest }) => {
  return (
    <label className='checkbox'>
      <input type={type} onChange={e => onChange(e.target.checked)} {...rest} />
      <span className='checkbox__icon'><ICheck/></span>

      <span className='checkbox__label'>
        {children}
      </span>
    </label>
  )
}

export default Checkbox
