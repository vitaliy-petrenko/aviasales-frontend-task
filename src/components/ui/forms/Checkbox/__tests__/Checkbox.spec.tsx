import React from 'react'
import Checkbox from '../index'
import { shallow } from 'enzyme'

describe('Checkbox component', () => {
  const
    component = shallow(<Checkbox/>),
    findInput = () => component.find({ type: 'checkbox' })

  it('renders properly', () => {
    expect(component).toMatchSnapshot()
  })

  it('has input', () => {
    expect(findInput()).toHaveLength(1)
  })

  it('can be unchecked', () => {
    expect(findInput().props().checked).toBe(false)
  })

  it('can be checked', () => {
    component.setProps({ checked: true })

    expect(findInput().props().checked).toBe(true)
  })

  it('calls onChange', () => {
    const onChange = jest.fn()

    component.setProps({ onChange })
    findInput().simulate('change')
    expect(onChange).toHaveBeenCalledTimes(1)
  })
})
