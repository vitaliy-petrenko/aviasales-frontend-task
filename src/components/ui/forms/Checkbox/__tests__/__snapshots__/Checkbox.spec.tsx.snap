// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component renders properly 1`] = `
<label
  className="checkbox"
>
  <input
    checked={false}
    type="checkbox"
  />
  <span
    className="checkbox__icon"
  >
    <ForwardRef(SvgCheck) />
  </span>
  <span
    className="checkbox__label"
  />
</label>
`;
